File: Swap_and_Range_Sum.py
import sys
input = lambda: sys.stdin.readline().strip()

class Tree:
    def __init__(self, n):
        self.tree = [0] * (n + 10)

    def lowbit(self, x):
        return x & -x

    def update(self, i, x):
        while i < len(self.tree):
            self.tree[i] += x
            i += self.lowbit(i)

    def query(self, i):
        ans = 0
        while i:
            ans += self.tree[i]
            i -= self.lowbit(i)
        return ans

    def sum(self, left, right):
        return self.query(right) - self.query(left)

n, q = map(int, input().split())
a = [0] + list(map(int, input().split()))
tree = Tree(n + 1)
for i in range(1, n + 1):
    tree.update(i, a[i])
for _ in range(q):
    op = list(map(int, input().split()))
    if op[0] == 1:
        tree.update(op[1], a[op[1] + 1] - a[op[1]])
        tree.update(op[1] + 1, a[op[1]] - a[op[1] + 1])
        a[op[1]], a[op[1] + 1] = a[op[1] + 1], a[op[1]]
    else:
        print(tree.sum(op[1] - 1, op[2]))

File: Swap_and_Range_Sum.cpp
#include <bits/stdc++.h>
using namespace std;

#define endl '\n'
#define int long long

struct Tree {
    vector<int> tree;
    
    Tree() {}
    Tree(int n) {
        tree.resize(n + 10);
    }
    
    int lowbit(int x) {
        return x & -x;
    }
    
    void update(int i, int x) {
        while (i <= tree.size()) {
            tree[i] += x;
            i += lowbit(i);
        }
    }
    
    int query(int i) {
        int ans = 0;
        while (i) {
            ans += tree[i];
            i -= lowbit(i);
        }
        return ans;
    }

    int sum(int left, int right) {
        return query(right) - query(left);
    }
};

void solve() {
    int n, q;
    cin >> n >> q;
    vector<int> a(n + 1);
    Tree tree(n + 1);
    for (int i = 1; i <= n; ++i) {
        cin >> a[i];
        tree.update(i, a[i]);
    }
    while (q--) {
        int op;
        cin >> op;
        if (op == 1) {
            int x;
            cin >> x;
            tree.update(x, a[x + 1] - a[x]);
            tree.update(x + 1, a[x] - a[x + 1]);
            swap(a[x], a[x + 1]);
        } else {
            int l, r;
            cin >> l >> r;
            cout << tree.sum(l - 1, r) << endl;
        }
    }
}

signed main() {
    cin.tie(nullptr)->sync_with_stdio(false);
    cout << fixed << setprecision(15);
    solve();
    return 0;
}
